darrencocco/moodle-tool_messagebroker
# Implement durable message processor scheduled task

# Testing

**NB**: Testing instructions assume a fresh install

1. Run the durable_message_processor task: `php admin/cli/scheduled_task.php --execute="\\tool_messagebroker\\task\\durable_message_processor"`
2. **Verify** the output looks like:
```
Execute scheduled task: durable_message_processor (tool_messagebroker\task\durable_message_processor)
... started 18:03:45. Current memory use 14.0 MB.
... used 2 dbqueries
... used 0.011904001235962 seconds
Scheduled task complete: durable_message_processor (tool_messagebroker\task\durable_message_processor)
```
3. Put the following script in to the root of any plugin and call it messagebroker.php:
```
<?php

class receiver implements tool_messagebroker\receiver\message_receiver {
    private string $topic;
    private string $style;
    private string $willsucceed;

    public function __construct(string $topic, string $style, bool $willsucceed) {
        $this->topic = $topic;
        $this->style = $style;
        $this->willsucceed = $willsucceed;
    }

    public static function instance(): self {
        return new self();
    }

    public function process_message(tool_messagebroker\message\immutable_message $message): bool {
        return $this->willsucceed;
    }

    public function get_preferred_message_processing_method(): string {
        return $this->style;
    }

    public function get_registered_topic(): string {
        return "/" . $this->topic . "/";
    }
}

function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true)
    ];
}
```
4. Purge caches
5. Set up web services (ensure the role assigned to the web service user has the capability `tool/messagebroker:submitmessage`
6. Make a curl request: `curl 'http://localhost/m/stable_401/webservice/rest/server.php?moodlewsrestformat=json' --data 'wsfunction=mbconnector_webservice_submit_message&topic=cool.movies&body=%7B%22some%22%3A%22thing%22%7D&wstoken=WS_TOKEN_HERE'`
7. **Verify** the response is `{"success":true,"stored":false,"receivers":3}`
8. Inspect the database and **verify** there are no records in the mdl_mbdatastore_standarddb_msg table
9. Alter the `build_message_receivers` function to look like:
```
function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::DURABLE, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true)
    ];
}
```
10. Make the curl request again
11. **Verify** the response is: `{"success":true,"stored":true,"receivers":3}`
12. **Verify** there is now a message stored in the DB
13. Execute the scheduled task again
14. **Verify** it completes and inspect the DB
15. **Verify** the completed flag is set to 1 and the processing time has been updated
16. Make the curl request 10 times to get 10 messages in to the system
17. Inspect the DB and **verify** a 10 new message were added
18. Run the scheduled task again
19. Inspect the DB and **verify** only one of the new tasks has had its completed flag set to 1
20. Browse to "Site administration" > "Plugins" > "Admin tools" > "Message Broker"
21. Set  "Messages per task" to 2
22. Run the scheduled task
23. Inspect the DB and **verify** two more of the messages have been marked as completed
24. On the plugin settings page, set "Message receiving mode" to "Ephemeral"
25. Run the scheduled task
26. Inspect the DB and **verify** that two more messages have the completed flag set to 1 (the processing style was set to ephemeral, which means there are no durable processors to process it - implying no errors with the processing)
27. On the plugin settings page, set the processing style to "Durable"
28. Alter the `build_message_receivers` function to look like:
```
function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, false),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true)
    ];
}
```
29.. Run the scheduled task and **verify** the output looks like (specific IDs may vary):
```
Execute scheduled task: durable_message_processor (tool_messagebroker\task\durable_message_processor)
... started 12:11:57. Current memory use 13.2 MB.
        Processing of message with ID 6 is incomplete. The following receivers still require it:
                receiver:172@/[PLUGIN]/messagebroker.php
        Processing of message with ID 5 is incomplete. The following receivers still require it:
                receiver:172@/[PLUGIN]/messagebroker.php
... used 3 dbqueries
... used 0.010945081710815 seconds
Scheduled task complete: durable_message_processor (tool_messagebroker\task\durable_message_processor)
```
30. Inspect the DB and **verify** the completed flag of the new remaining messages is still set to 0 but the processing time has been updated
31. Update `build_message_receivers` - changing the false back to true
32. Run the scheduled task again
33. Inspect the DB and **verify** that two more messages have the completed flag set to 1
34. Alter the `build_message_receivers` function to look like:
```
function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('dvds', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
    ];
}
```
35. Run the scheduled task
36. Inspect the DB and **verify** that two more messages have the completed flag set to 1 (we altered the receivers to be uninterested in this message, which means there are no durable processors to process it - implying no errors with the processing)

Fixes #5
